Support for resource names

michaelbausor opened the pull request:
This PR adds support for typed resource name objects to toolkit.

garrettjonesgoogle commented on src/test/java/com/google/api/codegen/testdata/library.proto:
```diff
@@ -14,12 +14,22 @@ import "google/protobuf/timestamp.proto";
 
+option (google.gapic.format) = {
+  format_name: 'ShelfName',
+  format_string: 'bookShelves/{shelf_id}/'
```
Do we want the trailing slash?

michaelbausor replied:
No, will fix

garrettjonesgoogle commented on src/test/java/com/google/api/codegen/testdata/java_test_library.baseline:
```diff
@@ -626,7 +628,7 @@ public class LibraryServiceTest {
     mockLibraryService.setResponses(expectedResponses);
 
-    String formattedName = LibraryServiceApi.formatBookName("[SHELF]", "[BOOK]");
+    BookName formattedName = BookName.create("[SHELF]", "[BOOK]");
```
It occurs to me that we could drop the addition of 'formatted' to parameter names since we have the resource types now

michaelbausor replied:
Should we drop it in all cases? Or just when a resource type is used?

garrettjonesgoogle replied:
Oh, I guess just when a resource is used.

garrettjonesgoogle commented on src/test/java/com/google/api/codegen/testdata/library.proto:
```diff
@@ -278,7 +288,7 @@ message CreateBookRequest {
 message PublishSeriesRequest {
   // The shelf in which the series is created.
-  Shelf shelf = 1;
+  Shelf shelf = 1 [(google.gapic.format_name) = "ShelfName"];
```
This is a shelf object, it shouldn't have a format

garrettjonesgoogle commented on src/main/resources/com/google/api/codegen/java/method_sample.snip:
```diff
@@ -76,56 +77,4 @@
-      $unhandledCase: {@line.lineType.toString}$
-    @end
-  @end
-@end
```
Wait how do we get away with not having this?

michaelbausor replied:
Added import of initcode.snip

garrettjonesgoogle commented on src/main/resources/com/google/api/codegen/java/initcode.snip:
```diff
@@ -47,7 +47,9 @@
     $unhandledCase: {@initValue.type}$
   @end
-@end
+@end
```
newline

garrettjonesgoogle commented on src/main/java/com/google/api/codegen/util/ResourceNameUtil.java:
```diff
@@ -0,0 +1,54 @@
+  }
+
+  public static boolean hasResourceName(Field field) {
+    // TODO(michaelbausor): do this check in a less horrific way
```
i love this comment

garrettjonesgoogle commented on src/main/java/com/google/api/codegen/metacode/InitCodeNode.java:
```diff
@@ -94,6 +95,13 @@ public TypeRef getType() {
   /*
+   * Get the TypeRef of the node.
+   */
+  public Field getField() {
```
wrong documentation

michaelbausor commented:
Updated and fixed issues. Major outstanding TODO is to get the resource_name_format.proto into googleapis (https://github.com/googleapis/googleapis/pull/152) then regenerate grpc-core-protos maven package and take a dependency on that, instead of including the resource_name_format.proto in toolkit.

garrettjonesgoogle commented on src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java:
```diff
@@ -97,7 +95,8 @@ private SurfaceTransformerContext getSurfaceTransformerContextFromService(Interf
         modelTypeTable,
-        new NodeJSSurfaceNamer(getApiConfig().getPackageName()));
+        new NodeJSSurfaceNamer(getApiConfig().getPackageName()),
+        false);
```
Ok I'm about to dump some work on you here...

It's hard to know what this "false" means in code like this. In general I'd like to avoid booleans as function parameters.

At a higher level, I think we might have a developing case for something like a FeatureConfig set of classes. Each language would have to provide a subclass (similar to SurfaceNamer). We should definitely put your new setting into it (resourceNameTypesEnabled), and any other globally-applicable features in the future would go in there too.

garrettjonesgoogle commented on src/main/java/com/google/api/codegen/transformer/ApiMethodTransformer.java:
```diff
@@ -409,8 +415,20 @@ private RequestObjectParamView generateRequestObjectParam(
+        context.resourceNameTypesEnabled() && ResourceNameUtil.hasResourceName(field);
+
+    //System.out.println(
+    //    "generateRequestObjectParam useResourceNameFormatOptionVar " + useResourceNameFormatOption);
```
delete?

garrettjonesgoogle commented on src/main/java/com/google/api/codegen/transformer/InitCodeTransformer.java:
```diff
@@ -235,8 +251,16 @@ private InitCodeView buildInitCodeView(
+      for (Field field : context.getMethod().getInputType().getMessageType().getFields()) {
+        if (ResourceNameUtil.hasResourceName(field)) {}
+      }
+    }
```
Why this no-op loop?

garrettjonesgoogle commented on src/main/java/com/google/api/codegen/transformer/InitCodeTransformer.java:
```diff
@@ -376,25 +417,69 @@ private InitValueView getInitValue(
-          namer.getVariableName(item.getIdentifier(), item.getInitValueConfig()));
+
+      boolean useResourceNameFormatOption =
+          context.resourceNameTypesEnabled() && ResourceNameUtil.hasResourceName(item.getField());
```
I am seeing this pattern repeated a lot

garrettjonesgoogle commented on src/main/java/com/google/api/codegen/transformer/SurfaceNamer.java:
```diff
@@ -121,16 +119,22 @@ public String getApiSettingsBuilderVarName(Interface interfaze) {
+  /** The variable name for the given identifier. */
+  public String getVariableName(Name identifier) {
+    return varName(identifier);
+  }
```
This function is unnecessary, varName is public.

michaelbausor commented:
Added FeatureConfig class, addressed PR comments. List support still TODO

garrettjonesgoogle commented on src/main/java/com/google/api/codegen/transformer/MethodTransformerContext.java:
```diff
@@ -73,6 +75,11 @@ public CollectionConfig getCollectionConfig(String entityName) {
+  public boolean useResourceNameFormatOption(InitCodeNode item) {
+    return getFeatureConfig().resourceNameTypesEnabled()
+        && ResourceNameUtil.hasResourceName(item.getField());
+  }
```
I think it would be more appropriate to put this in FeatureConfig.

garrettjonesgoogle commented on src/main/java/com/google/api/codegen/transformer/php/PhpModelTypeNameConverter.java:
```diff
@@ -175,4 +175,10 @@ public String renderPrimitiveValue(TypeRef type, String value) {
+  @Override
+  public TypeName getTypeNameForTypedResourceName(
+      ProtoElement elem, String typedResourceShortName) {
+    throw new UnsupportedOperationException("getTypeNameForTypedResourceName not supported by PHP");
```
Why implement this? It duplicates the value in FeatureConfig.

michaelbausor replied:
It is part of the ModelTypeNameConverter interface, so has to be implemented

garrettjonesgoogle replied:
oh, that makes sense. nevermind.

garrettjonesgoogle commented on src/main/resources/com/google/api/codegen/java/method_sample.snip:
```diff
@@ -76,56 +77,4 @@
-    $unhandledCase: {@initValue.type}$
-  @end
-@end
+@end
```
newline

michaelbausor commented:
Added list and page streaming support, and an additional test method in library.proto, PTAL

garrettjonesgoogle commented on src/main/java/com/google/api/codegen/transformer/InitCodeTransformer.java:
```diff
@@ -342,61 +370,103 @@ private InitCodeLineView generateMapInitCodeLine(
+    List<String> formatFunctionArgs = new ArrayList<>();
+    for (String entityName : varList) {
+      String entityValue =
+          "\"[" + LanguageUtil.lowerUnderscoreToUpperUnderscore(entityName) + "]\"";
```
Switch to use the Name class

michaelbausor replied:
I will use Name class on the entityName. Currently the Name class doesn't support special characters (like " and [). Should I add support for this, or do we want to keep the Name class only applicable to characters that are valid in normal names?

garrettjonesgoogle replied:
I specifically didn't want to reference LanguageUtil. You can leave the special characters out of it for now.

michaelbausor commented:
Merged the C# changes - will wait and merge the Go changes when available

michaelbausor commented:
Merged Go MVVM changes, PTAL

garrettjonesgoogle commented on src/main/java/com/google/api/codegen/transformer/FeatureConfig.java:
```diff
@@ -14,8 +14,19 @@
+
 public class FeatureConfig {
 
+  public boolean resourceNameTypesEnabled() {
```
Add docs to the methods in this class
